### ISSUE TITLE
Idea: move mimir go client into own package

### ISSUE THREAD

**Issue by Syndlex** [grafana/mimir#10315]
### What is the problem you are trying to solve?

help opensource community use mimir's api to build IaC tools 

### Which solution do you envision (roughly)?

Extract everything that is in: https://github.com/grafana/mimir/tree/main/pkg/mimirtool/client into its on go package.

### Have you considered any alternatives?

writing everything by myself

### Any additional context to share?

I think it's interesting that you have this code in multiple repositories yourself.
You can find it in Grizzly, Mimir Tools, and I think I found another repository with the same code.

### How long do you think this would take to be developed?

Small (<= 1 month dev)

**Comment by pracucci** [comment]
I'm not against the proposal, but what's the problem directly importing `pkg/mimirtool/client`? Why does moving it to a different package will solve the issue?

**Comment by 56quarters** [comment]
I am not a fan of moving the client outside this repository and introducing another package with APIs and the obligations that come with that. Something like this was proposed before here https://github.com/grafana/mimir/issues/6677. There shouldn't be anything preventing the client from being imported as-is like Marco mentioned besides the number of dependencies that Mimir pulls in.

**Comment by Syndlex** [comment]
I think moving it out off this repo would be a bad idea too. 
Moving it into its own path/package would have some benefits. 
Something like `PKG/client` or something. 

These are the benefits I see:
* Clearer communication that there is a implementation for a client.
  * For me this is the main benefit. Enable the community to build stuff for the mimir system without implementing boilerplate API code.
* When I loaded the mimirtools package it was very bloated (maybe a mistake on my side).
* There is also the dry argument but I always think this is not the best argument.
  * Currently it feels like grizzly alloy don't know that there could use this client.

**Comment by Syndlex** [comment]
Yes Tested it again to load it into my path and it loaded the entire mimir codebase.  
```
go get github.com/grafana/mimir/pkg/mimirtool/client
 go: downloading github.com/grafana/mimir v0.0.0-20250103212259-2a4ba9bc9a77 
 go: downloading github.com/thanos-io/objstore v0.0.0-20241226233006-a14268e0e939 
 go: downloading google.golang.org/genproto/googleapis/rpc v0.0.0-20241230172942-26aa7a208def
 go: downloading go.etcd.io/etcd/api/v3 v3.5.16
 go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.16
 go: downloading go.etcd.io/etcd/client/v3 v3.5.16
```

**Comment by colega** [comment]
I feel there's some miscommunication in this conversation.

Just to clarify: I understand that we're talking about creating a `go.mod` file in the `pkg/mimirtool/client` folder, right?

I think that's a sensible idea.

**Comment by Syndlex** [comment]
Basically yes!

I think there are some missing mimir apis to be a full client implementation but better than nothing :).

**Comment by colega** [comment]
Would you like to send a PR for this?

**Comment by Syndlex** [comment]
Hey,

yes I can try but I think my Go Skills are not enough,

But I don't know if this is working. 
Without reordering some File.

`pkg/mimirtool/client/rules.go` is importing `github.com/grafana/mimir/pkg/mimirtool/rules/rwrulefmt` 
this would lead to a cyclic dependencie between the `go.mod` root and the `go.mod` `pkg/mimirtool/client`

With just adding a go.mod into the `pkg/mimirtool/client` directory i get this with `go mod tidy`:

```
 found package github.com/grafana/mimir/pkg/mimirtool/client in multiple modules:
        github.com/grafana/mimir v0.0.0-20250315192811-c301d4361fd2 (/home/mfeix@corp.exxcellent.de/go/pkg/mod/github.com/grafana/mimir@v0.0.0-20250315192811-c301d4361fd2/pkg/mimirtool/client)
        github.com/grafana/mimir/pkg/mimirtool/client (/home/mfeix@corp.exxcellent.de/Repo/gilde/cloud/mimir/pkg/mimirtool/client)
```

I tried to write. `replace github.com/grafana/mimir/pkg/mimirtools/client v0.0.0 => ./pkg/mimirtool/client` into root `go.mod`

But this didn't work and gave me this output in the root `go.mod:
```
go: finding module for package github.com/grafana/mimir/pkg/mimirtool/client
go: github.com/grafana/mimir/pkg/mimirtool/commands imports
        github.com/grafana/mimir/pkg/mimirtool/client: no matching versions for query "latest"
```

Do you have any Idear?
